packages/core/src/utils/multipleTemplates.ts

```diff
@@ -0,0 +1,178 @@
+  )
+  return Boolean(foundRewrite)
+}
+
+/**
+ * Find the best PLP template from the CMS based on the slug passed as param.
+ *
+ * This function iterates the slug until there is no slashes (/), prioritizing the following order:
+ * 1. A PLP template that matches the subcategory (e.g. slug = /department/category/subcategory).
+ * 2. A PLP template that matches the category (e.g. slug = /department/category).
+ * 3. A PLP template that matches the department (e.g. slug = /department).
+ * 4. If no matches are found, use the generic PLP template.
+ *
+ * @param pages
+ * @param originalSlug
+ * @returns The best PLP template page for the slug
+ */
+export function findBestPLPTemplate(
+  pages: Partial<PLPContentType>[],
+  originalSlug: string,
+  rewrites: Rewrite[] | RewritesConfig
+) {
+  let slug = normalizePLPSlug(originalSlug)
+  let foundPageTemplate = findPLPTemplateBySlug(pages, slug, rewrites)
+
+  while (!foundPageTemplate && slug.lastIndexOf('/') !== -1) {
+    slug = slug.substring(0, slug.lastIndexOf('/'))
+    foundPageTemplate = findPLPTemplateBySlug(pages, slug, rewrites)
+  }
+
+  return (
+    foundPageTemplate ||
+    pages.find((page) => !page.settings?.template?.value) ||
+    pages[0]
+  )
+}
+
+export function normalizePDPTemplate(templateValue: string) {
+  // Remove extra slashes, white spaces at the beginning and end
+  let formattedValue = templateValue.trim().replace(/^\/+|\/+$/g, '')
+
+  // Add a slash at the beginning if not present
+  if (!formattedValue.startsWith('/')) {
+    formattedValue = '/' + formattedValue
+  }
+
+  // Add a slash at the end if not ending with '/p'
+  if (!formattedValue.endsWith('/p')) {
+    formattedValue += '/'
+  }
+
+  // Remove duplicate slashes and white spaces throughout the string
+  formattedValue = formattedValue.replace(/\/+/g, '/').replace(/\s+/g, '')
+
+  return formattedValue.toLowerCase()
+}
+
+/**
+ * Find the best PDP template from the CMS based on the slug or in the product category tree.
+ * Prioritizing the following order:
+ *
+ * 1. A PDP template that matches the page slug (e.g. slug = /apple-magic-mouse/p).
+ * 2. A PDP template that matches the product subcategory (e.g. /department/category/subcategory).
+ * 3. A PDP template that matches the product category (e.g. /department/category).
+ * 4. A PDP template that matches the product department (e.g. /department).
+ * 5. If no matches are found, use the generic PDP template.
+ *
+ * @param pages
+ * @param originalSlug
+ * @param product
+ * @returns The best PDP template page for the slug
+ */
+export function findBestPDPTemplate(
+  pages: Partial<PDPContentType>[],
+  slug: string,
+  product: ServerProductQueryQuery['product']
+) {
+  // productSlugAndCategoryTree with the prioritized order. [slug, subcategory tree, category tree, department]
+  const productSlugAndCategoryTree = product?.breadcrumbList?.itemListElement
+    ? [...product?.breadcrumbList?.itemListElement]
```

Comment:
itemListElement already a list, not?

Reply:
right, but if we mutate this original list, this can affect the breadcrumb component (and maybe others). Thus, we create a new list.